editoast/fga_migrations/src/migration.rs

```diff
@@ -0,0 +1,509 @@
+            },
+            telemetry: TelemetrySettings {
+                endpoint: url::Url::parse(&client.telemetry_endpoint).expect("Invalid url"),
+            },
+        }
+    }
+}
+
+#[derive(fga::Type, fga::User, derive_more::FromStr, Debug, Clone, PartialEq, Eq, Hash)]
+struct Migration(pub usize);
+
+#[derive(fga::Type, fga::Object, derive_more::FromStr, Debug, Clone, PartialEq, Eq, Hash)]
+struct Model(pub String);
+
+fga::relations! {
+    Model {
+        apply: Migration
+    }
+}
+
+#[derive(Debug, thiserror::Error)]
+pub enum MigrationError {
+    #[error("Invalid migration name `{0}`")]
+    InvalidMigrationName(String),
+    #[error(transparent)]
+    InitializationError(#[from] InitializationError),
+    #[error(transparent)]
+    QueryError(#[from] QueryError),
+    #[error(transparent)]
+    DeleteTuplesError(#[from] Either<RequestFailure, TooManyTuples>),
+    #[error(transparent)]
+    RequestFailure(#[from] RequestFailure),
+}
+
+#[instrument(name = "migration-run")]
+pub async fn run_migrations(
+    settings: Settings,
+    target_migration: TargetMigration,
+) -> anyhow::Result<Client, MigrationError> {
+    // Retrieve the new DSL model definition
+    let migration_number = match target_migration {
+        TargetMigration::Latest => OPENFGA_MIGRATIONS.len() - 1,
+        TargetMigration::Name(migration_name) => OPENFGA_MIGRATIONS
+            .get_index_of(&migration_name.as_str())
+            .ok_or(MigrationError::InvalidMigrationName(
+                migration_name.to_string(),
+            ))?,
+    };
+    let migration_model_dsl = OPENFGA_MIGRATIONS[migration_number];
+    info!("Retrieved migration DSL model: {}", migration_model_dsl);
+
+    // Get clients to the authorization and migration stores, create the stores if needed
+    let mut client_authorization_store = Client::try_store_or_create(
+        &settings.migration_settings.authorization_store,
+        settings.migration_settings.connection_settings.clone(),
+        None,
+    )
+    .await?;
+    let migrations_store_model = compile_model(MIGRATIONS_MODEL);
+    let client_migrations_store = Client::try_store_or_create(
+        &settings.migration_settings.migration_store,
+        settings.migration_settings.connection_settings.clone(),
+        Some(migrations_store_model),
+    )
+    .await?;
+    info!(
+        store.migrations.id = client_migrations_store.store.id,
+        store.authz.id = client_authorization_store.store.id,
+        "Retrieved or created migration and authorization stores"
+    );
+
+    // Make the hash of the new DSL model
+    let mut hasher = Sha1::new();
+    hasher.update(migration_model_dsl);
+    let new_model_hash = hasher.finalize();
+    info!(migration.model.hash = format!("{new_model_hash:x}"));
+    let new_model_hash = format!("{new_model_hash:x}");
+
+    let migration_model = Model(new_model_hash);
+
+    // Remove existing migration tuples or exit if the target migration is already applied
```

Comment:
Reading this, I'm doubting that the migration store model is right. Correct me if I'm wrong, the only information we need to store is "migration 3 is applied". So the whole algorithm to apply migration 4 should be:

* which migration is applied?
* is it migration 4? If yes all good
* if not, remove the tuple and write "migration 4 is applied"

I believe this complexity comes from de-correlating migration indexes and hashes. Couldn't migration objects be like `migration:4-<hash>`?